### PR TITLE
Pluggable exchange

### PR DESCRIPTION
**Pull request by lzach**
Based on discussions with @BertR  and @perllaghu, I've created a pluggable version of the exchange, to allow custom exchanges for people who don't want to/can't use the default one. See #1130 for a bit more information.

The exchange file-tree has changed as follows:

<pre>
exchange/
├── abc/
│   ├── collect.py
│   ├── ...
│   └── submit.py
├── default/
│   ├── collect.py
│   ├── ...
│   └── submit.py
└── exchange_factory.py
</pre>

Here abc/ contains  abstract classes for each different exchange service (they aren't literally abstract, as the ABC meta-class conflicts with the LoggingConfigurable meta-class). The default/ folder contains the default implementations for each exchange service, which is just the old implementation.

The exchange_factory.py file contains the ExchangeFactory class, which is the interface for creating pluggable exchange services. It uses the Type traitlet for class variables for each exchange service, which is set to be a subclass of the corresponding class in abc/ and defaults to the corresponding class in  default/.

The ExchangeFactory class has instance methods that instantiate each of the exchange services. The new way of constructing the exchange classes would look something like:

<pre>
# lister = ExchangeLister(...) # old way
exchange = ExchangeFactory(config=config) # or parent=self, or similar
lister = exchange.List(...)
</pre>

For the apps, I've created an instance variable in the NbGrader app for the ExchangeFactory class, so that subclasses can use self.exchange.<exchange service> to create the services. The ExchangeFactory class gets initiated and configured in NbGrader.start, so before that method has been called, subclasses can't access the self.exchange variable. (Note, I removed line 39 in dbapp.py as it was creating a ExchangeList object and assigning it to an unused variable. Since this was done before the call to NbGrader.start it couldn't use the self.exchange, and as it was unused anyway, I figured I'd just delete the line.)

I also updated server_extensions/assignment_list/handlers.py to use the new pluggable exchange.

Finally, I should note that the method names in the ExchangeFactory class do not conform to PEP8. I did this to keep the names as close to the previous names as possible. If this is an issue, it should not be too much of a problem to change the names at this point.

Hope this all made sense.

### CODE REVIEW

**Comment by jhamrick** [comment]
Thanks a lot, this is super exciting! Sorry I've been slow to get to reviews; I will try to find some time to dig through this in the next few days.

File: nbgrader/apps/api.py [hunk 1]
```diff
@@ -6,12 +6,12 @@
...
 import warnings
 
-from traitlets.config import LoggingConfigurable, Config
+from traitlets.config import LoggingConfigurable, Config, get_config
 from traitlets import Instance, Enum, Unicode, observe
 
 from ..coursedir import CourseDirectory
 from ..converters import GenerateAssignment, Autograde, GenerateFeedback
-from ..exchange import ExchangeList, ExchangeReleaseAssignment, ExchangeReleaseFeedback, ExchangeFetchFeedback, ExchangeCollect, ExchangeError, ExchangeSubmit
+from ..exchange import ExchangeFactory, ExchangeError # ExchangeList, ExchangeReleaseAssignment, ExchangeReleaseFeedback, ExchangeFetchFeedback, ExchangeCollect, ExchangeError, ExchangeSubmit
```
Comment (jhamrick) [inline nbgrader/apps/api.py]: Could you remove the commented code here?

File: nbgrader/exchange/abc/exchange.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+from nbgrader.utils import check_directory, ignore_patterns, self_owned
+from nbgrader.coursedir import CourseDirectory
+from nbgrader.auth import Authenticator
+
+
+class ExchangeError(Exception):
+    pass
+
+
+class Exchange(LoggingConfigurable):
```
Comment (jhamrick) [inline nbgrader/exchange/abc/exchange.py]: It seems like some of the traits and methods of this class make some strong assumptions about the files being located in a unix-like filesystem (e.g. the 'root') traitlet. Perhaps it would better to move these to the default exchange implementation rather than having them in the base class?

File: pytest.ini [hunk 3]
```diff
@@ -1,4 +1,4 @@
 [pytest]
-addopts = --strict-markers
+;addopts = --strict-markers
```
Comment (jhamrick) [inline pytest.ini]: Looks like this unrelated change snuck in here?

**Comment by jhamrick** [comment]
Oh actually I realized the tests do not explicitly create the exchange objects, so no work actually to do there!

**Comment by lzach** [comment]
For some reason the travis build for nbextensions fails, even though it passes on my machine (both are using python 3.6). @BertR is checking whether it passes on his machine to determine whether it's an issue with my setup or something else.

**Comment by lzach** [comment]
It appears that there's something different with my setup... trying to figure out what.

**Comment by lzach** [comment]
Turns out that it was some changes that had been pushed to master between me pushing the original and pushing the requested changes that broke the tests. Fixed now.

File: nbgrader/exchange/abc/exchange.py (hunk 2 above)
Comment (jhamrick) [reply nbgrader/exchange/abc/exchange.py]: Yes, that's a good point. An easy way to get around this would be to list the options for the default exchange classes in the `NbGrader` app. Specifically, in `apps/baseapp.py`, you can have:

```
        # include all the exchange actions
        for ex_name in exchange.default.__all__:
            ex = getattr(exchange.default, ex_name)
            if hasattr(ex, "class_traits") and ex.class_traits(config=True):
                classes.append(ex)
```

on line 153 (i.e. instead of importing the classes from `exchange` you import them from `exchange.default`).

This will of course though not help if you are using a different exchange than the default one. But in that case, perhaps we can assume that since it is a custom thing you would be familiar with the available options?

**Comment by lzach** [comment]
I've added in the code to pull the configuration from the default package (it also takes from the abc package, as there are some attributes that ought to be common to all implementations). I also added some documentation for how to write exchange plugins. It was originally written by @perllaghu, but I added in the new bits. Let me know if you want any changes to this.

The windows test fail for some reason and I can't check why as I don't have a windows machine.

**Comment by moorepants** [comment]
Hi folks, a student group working with me has created ngshare:

https://github.com/lxylxy123456/ngshare

This allows nbgrader to be used on clusters and they've shown a very nice demo. But, it relies on this pull request. They have about 4 more weeks before school lets out and can spend some resources on getting this PR merged.

What can we do to help here?

I'm cc/ing the students: @Lawrence37, @lxylxy123456, @rkevin-arch, @aalmanza1998.

**Comment by perllaghu** [comment]
> Hi folks, a student group working with me has created ngshare:
> 
> https://github.com/lxylxy123456/ngshare
> 
> This allows nbgrader to be used on clusters and they've shown a very nice demo. But, it relies on this pull request. They have about 4 more weeks before school lets out and can spend some resources on getting this PR merged.
> 
> What can we do to help here?
> 
> I'm cc/ing the students: @Lawrence37, @lxylxy123456, @rkevin-arch, @aalmanza1998.

It's failing on python 3.5 & 3.6 builds (possibly because we've used `f strings` a lot) - which then fails the azure.pipeline.

It would be interesting to see what else fails under 3.5 & 3.6

.... but, other than that - we're just waiting on a maintainer to do a final review & merge.

**Comment by moorepants** [comment]
@jhamrick Are you the only maintainer that can approve merges? What can we do to get this one through the pipeline?

> It's failing on python 3.5 & 3.6 builds (possibly because we've used f strings a lot) - which then fails the azure.pipeline.

My guess, at this point in time, they will want to support Python 3.6+, so the fstrings will likely have to be removed.

**Comment by moorepants** [comment]
@Lawrence37, @lxylxy123456, @rkevin-arch, @aalmanza1998 I think it would be worth it if you all reviewed this PR. Having more eyes on it could help @jhamrick feel comfortable to merge. Will you all review?

File: nbgrader/apps/api.py [hunk 4]
```diff
@@ -48,7 +49,7 @@ def _log_level_changed(self, change):
             self.log_level = new
         self.log.setLevel(new)
 
-    def __init__(self, coursedir=None, authenticator=None, **kwargs):
+    def __init__(self, coursedir=None, authenticator=None, exchange=None, **kwargs):
```
Comment (moorepants) [inline nbgrader/apps/api.py:52]: Needs an addition to the docstring explaining the new kwarg.

File: nbgrader/exchange/default/fetch_feedback.py [hunk 5]
```diff
@@ -2,11 +2,13 @@
 import shutil
 import glob
 
-from .exchange import Exchange
-from ..utils import check_mode, notebook_hash, make_unique_key, get_username
+from nbgrader.exchange.abc import ExchangeFetchFeedback as ABCExchangeFetchFeedback
+from nbgrader.exchange.default import Exchange
```
Comment (moorepants) [inline nbgrader/exchange/default/fetch_feedback.py:6]: Why have all the relative imports been changed? They should work as was, no?

File: nbgrader/exchange/abc/list.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        pass
+
+    def format_inbound_assignment(self, info):
+        pass
+
+    def format_outbound_assignment(self, info):
+        pass
+
+    def parse_assignments(self):
+        pass
```
Comment (Lawrence37) [inline nbgrader/exchange/abc/list.py:30]: I would remove these functions.

File: nbgrader/exchange/abc/list.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+    def list_files(self):
+        pass
+
+    def remove_files(self):
+        pass
```
Comment (Lawrence37) [inline nbgrader/exchange/abc/list.py:36]: Perhaps implement the `start` function, which will ensure `inbound` and `cached` are not simultaneously set and call `list_files` or `remove_files` depending on the value of `remove`. Moving the `start` function from default.ExchangeList here would suffice.

I suggest documenting these two functions and making them raise a NotImplementedError, similar to `init_src` and `init_dest` in default.Exchange.

File: nbgrader/docs/source/exchange/exchange_api.rst [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+returns.... *nothing*
+
+When writing your own Exchange
+++++++++++++++++++++++++++++++
+
+- You to need to consider stopping students from seeing each others submissions
+- nbgrader functionality requires a file called ``timestamp.txt`` to be in the submission, containing the timestamp of that submission. The creation of this file is the responsibility of this class.
+- Whilst nothing is done *as yet*, the default exchange checks the names of submitted notebooks, and logs differences.
+- Submissions need to record ``student_id``, as well as ``course_id`` & ``assignment_id``
+- The default exchange copies files to both an ``inbound`` and ``cache`` store. This may be significant considering ``ExchangeList``
```
Comment (Lawrence37) [inline nbgrader/docs/source/exchange/exchange_api.rst:309]: Did the last sentence get cut off? If not, explain why it is significant.

File: nbgrader/exchange/abc/collect.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+
+    def init_src(self):
+        pass
+
+    def init_dest(self):
+        pass
+
+    def copy_files(self):
+        pass
```
[the remark this answers is not in the document]
Comment (moorepants) [reply nbgrader/exchange/abc/collect.py:26]: Maybe it's designed as an abstract base class?

[the remark this answers is not in the document]
Comment (moorepants) [reply nbgrader/exchange/abc/collect.py:26]: They could probably be removed though.

**Comment by moorepants** [comment]
@Lawrence37 Thanks! The feedback looks good.

**Comment by moorepants** [comment]
@lzach and @perllaghu Will you all have any cycles to update this PR? Or would you like us to open a new PR that builds on this? We could do either.

File: nbgrader/apps/dbapp.py [hunk 10]
```diff
@@ -36,7 +36,6 @@ class DbBaseApp(NbGrader):
 
     def start(self):
         if sys.platform != 'win32':
-            lister = ExchangeList(coursedir=self.coursedir, parent=self)
```
Comment (aalmanza1998) [inline nbgrader/apps/dbapp.py:39]: Why was 'from ..exchange import ExchangeList' left in the imports if this line was removed?

File: nbgrader/docs/source/exchange/exchange_plugin.rst [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    │   ├── collect.py
+    │   ├── ...
+    │   └── submit.py
+    ├── default/
+    │   ├── collect.py
+    │   ├── ...
+    │   └── submit.py
+    └── exchange_factory.py
+
+The ``exchange.abc`` package contain all the Abstract Base Classes that custom exchange packages need to implement.
```
Comment (aalmanza1998) [inline nbgrader/docs/source/exchange/exchange_plugin.rst:24]: small typo, should say **contains** instead of **contain**

File: nbgrader/exchange/abc/release_feedback.py [hunk 12]
```diff
@@ -0,0 +1,5 @@
+from .exchange import Exchange
+
+
+class ExchangeReleaseFeedback(Exchange):
+    pass
```
Comment (aalmanza1998) [inline nbgrader/exchange/abc/release_feedback.py:5]: missing newline at the end of file

File: nbgrader/exchange/default/__init__.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    "ExchangeCollect",
+    "ExchangeFetch",
+    "ExchangeFetchAssignment",
+    "ExchangeFetchFeedback",
+    "ExchangeList",
+    "ExchangeRelease",
+    "ExchangeReleaseAssignment",
+    "ExchangeReleaseFeedback",
+    "ExchangeSubmit"
+]
```
Comment (aalmanza1998) [inline nbgrader/exchange/default/__init__.py:24]: missing newline at the end of file

File: nbgrader/server_extensions/assignment_list/handlers.py [hunk 14]
```diff
@@ -14,9 +14,9 @@
 from traitlets.config import LoggingConfigurable, Config
 from jupyter_core.paths import jupyter_config_path
 
+from ...exchange import ExchangeFactory, ExchangeError
 from ...apps import NbGrader
 from ...coursedir import CourseDirectory
-from ...exchange import ExchangeList, ExchangeFetchAssignment, ExchangeFetchFeedback, ExchangeSubmit, ExchangeError
```
Comment (aalmanza1998) [inline nbgrader/server_extensions/assignment_list/handlers.py:19]: There is also an import for ExchangeList [here](https://github.com/lzach/nbgrader/blob/fe3ef73cdf22f79f8ca50ea4511366aca4e74ad3/nbgrader/server_extensions/formgrader/apihandlers.py#L8)

File: nbgrader/exchange/abc/submit.py [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+            "notebooks from the released assignment notebooks."
+        )
+    ).tag(config=True)
+
+    add_random_string = Bool(
+        True,
+        help=dedent(
+            "Whether to add a random string on the end of the submission."
+        )
+    ).tag(config=True)
```
[the remark this answers is not in the document]
Comment (perllaghu) [reply nbgrader/exchange/abc/submit.py:22]: The "Everyone has read access" of standard nbgrader is how the instructor [just another user on the system] can read submissions copied into the directory by a student  [just another user on the system]

File: nbgrader/apps/api.py (hunk 4 above)
Comment (lzach) [reply nbgrader/apps/api.py:52]: Fixed this on my local branch.

File: nbgrader/exchange/default/fetch_feedback.py (hunk 5 above)
Comment (lzach) [reply nbgrader/exchange/default/fetch_feedback.py:6]: It was a long time since I wrote this, so can't quite remember, but I assume it was just a side effect of changing the structure of the package. It shouldn't matter either way.

File: nbgrader/apps/api.py [hunk 16]
```diff
@@ -92,7 +102,7 @@ def __init__(self, coursedir=None, authenticator=None, **kwargs):
 
         else:
             self.course_id = ''
-            self.exchange = ''
+            self.exchange_root = ''
```
[the remark this answers is not in the document]
Comment (lzach) [reply nbgrader/apps/api.py:105]: Well spotted, I've fixed this on my local branch.

File: nbgrader/exchange/abc/collect.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lzach) [reply nbgrader/exchange/abc/collect.py:26]: I left some original structure in as I wanted feedback on what people thought would be useful to keep. These methods are definitely not needed though, and I've removed them in my local branch.

File: nbgrader/exchange/abc/fetch_assignment.py [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+from .exchange import Exchange
+from nbgrader.utils import check_mode
+
+
+class ExchangeFetchAssignment(Exchange):
+
+    replace_missing_files = Bool(False, help="Whether to replace missing files on fetch").tag(config=True)
+
+    def copy_if_missing(self, src, dest, ignore=None):
+        pass
```
[the remark this answers is not in the document]
Comment (lzach) [reply nbgrader/exchange/abc/fetch_assignment.py:15]: You're probably right, I've removed it from my local branch as it doesn't do anything anyway right now.

File: nbgrader/exchange/abc/list.py [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+import hashlib
+
+from traitlets import Bool
+from .exchange import Exchange
+
+
+def _checksum(path):
+    m = hashlib.md5()
+    m.update(open(path, 'rb').read())
+    return m.hexdigest()
```
[the remark this answers is not in the document]
Comment (lzach) [reply nbgrader/exchange/abc/list.py:11]: Removed from local branch.

File: nbgrader/exchange/default/fetch.py [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+class ExchangeFetch(ExchangeFetchAssignment, ABCExchangeFetch):
+
+    def __init__(self, *args, **kwargs):
+        super(ExchangeFetch, self).__init__(*args, **kwargs)
+        msg = (
+            "The ExchangeFetch class is now deprecated, please use "
+            "ExchangeFetchAssignment instead. This class will be removed in a "
+            "future version of nbgrader.")
+        warnings.warn(msg, DeprecationWarning)
+        self.log.warning(msg)
```
[the remark this answers is not in the document]
Comment (lzach) [reply nbgrader/exchange/default/fetch.py:16]: I've completely removed the default/fetch.py and default/release.py files, as they can't be instantiated using the factory anyway.

File: nbgrader/exchange/abc/submit.py [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+
+    def init_release(self):
+        pass
+
+    def check_filename_diff(self):
+        pass
```
[the remark this answers is not in the document]
Comment (lzach) [reply nbgrader/exchange/abc/submit.py:28]: You're probably right, I've removed them from my branch.

File: nbgrader/exchange/abc/submit.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (lzach) [reply nbgrader/exchange/abc/submit.py:22]: You're probably right that this is implementation specific and I've moved this option to default.ExchangeSubmit in my own branch.

File: nbgrader/exchange/abc/list.py (hunk 6 above)
Comment (lzach) [reply nbgrader/exchange/abc/list.py:30]: Removed from my local branch.

File: nbgrader/exchange/abc/list.py (hunk 7 above)
Comment (lzach) [reply nbgrader/exchange/abc/list.py:36]: Sounds reasonable and I've made this change in my local branch.

File: nbgrader/docs/source/exchange/exchange_api.rst (hunk 8 above)
Comment (lzach) [reply nbgrader/docs/source/exchange/exchange_api.rst:309]: I have no idea what happened to that sentence. I've removed it for now as I'm not sure what is supposed to be significant here.

File: nbgrader/apps/dbapp.py (hunk 10 above)
Comment (lzach) [reply nbgrader/apps/dbapp.py:39]: An oversight on my part, I've removed all unused imports I could find that relates to the exchange package now.

File: nbgrader/docs/source/exchange/exchange_plugin.rst (hunk 11 above)
Comment (lzach) [reply nbgrader/docs/source/exchange/exchange_plugin.rst:24]: Thanks, fixed.

File: nbgrader/exchange/abc/release_feedback.py (hunk 12 above)
Comment (lzach) [reply nbgrader/exchange/abc/release_feedback.py:5]: Fixed

File: nbgrader/exchange/default/__init__.py (hunk 13 above)
Comment (lzach) [reply nbgrader/exchange/default/__init__.py:24]: fixed.

File: nbgrader/server_extensions/assignment_list/handlers.py (hunk 14 above)
Comment (lzach) [reply nbgrader/server_extensions/assignment_list/handlers.py:19]: Removed.

**Comment by moorepants** [comment]
Thanks for the updates @lzach!

Does anyone know if @jhamrick is the only person with merge rights on this repo?

**Comment by lzach** [comment]
@moorepants If it's alright with you, I'll just push the fixes to this PR?

As far as I know, @jhamrick  is indeed the only one, though I'm not 100% sure.

**Comment by moorepants** [comment]
> If it's alright with you, I'll just push the fixes to this PR?

Definitely!

**Comment by moorepants** [comment]
We need to figure out the CI build failures too.

**Comment by moorepants** [comment]
Also, isn't this set of changes going to need unit tests? I didn't see any files labeled as tests on first glance.

**Comment by perllaghu** [comment]
Switch away from fstrings too

**Comment by lzach** [comment]
I'm currently writing tests for our version of the exchange, and will add in tests here as well. Almost all the code is covered by previous tests, as this mostly only changes the structure of the exchange package, whereas most of the logic remains exactly the same.

@perllaghu I can't find any f-strings in my nbgrader branch.  I'll run tests on all supported versions of python before I push, so hopefully I'll catch whatever is causing the checks to fail.

**Comment by perllaghu** [comment]
Ok.... Something else was making it fail the 3.5 & 3.6 environments
then....


But looking good!

**Comment by moorepants** [comment]
@Lawrence37, @lxylxy123456, @rkevin-arch, @aalmanza1998 You all can also help by looking into the CI errors.

File: nbgrader/exchange/abc/submit.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (perllaghu) [reply nbgrader/exchange/abc/submit.py:22]: If this is an nbgrader config option, we need to be sure the system doesn't break if a user has it defined in their config, and switch between the default file-based exchange and some external exchange.

File: nbgrader/docs/source/exchange/exchange_api.rst (hunk 8 above)
Comment (perllaghu) [reply nbgrader/docs/source/exchange/exchange_api.rst:309]: The `cached` flag is used in base `nbgrader` to get the list of files in the submission cache (see https://github.com/jupyter/nbgrader/blob/dd0d66a41685c911971b32cbc3243659b2813a39/nbgrader/exchange/list.py#L20)

File: nbgrader/exchange/abc/collect.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (perllaghu) [reply nbgrader/exchange/abc/collect.py:26]: If we follow the logic of the file-based exchange - all common methods are implemented in the base `Exchange` class, with `raise NotImplementedError`, and then defined in the actual implementation classes (with a `pass` if they do nothing.

This would be good login to follow if we want to make this code as usable as possible for others.

**Comment by lzach** [comment]
I've pushed the latest changes. All tests passes except for the mypy ones. This seems to a problem on all pull requests though, so not sure what can be done about it. Mypy doesn't seem to like the demos folder, as far as I can tell.

**Comment by lzach** [comment]
Pushed a fix for the failing mypy tests as described by @BertR in #1328

**Comment by willingc** [comment]
@jhamrick I would recommend dropping Python 3.5 now since it goes EOL in a few months and has been in security fix mode since last year. Dropping it would allow f-strings and better asyncio support.
